drivers/virtio/net/virtio_net.c

```diff
@@ -510,30 +520,41 @@ static int virtio_netdev_rxq_enqueue(struct virtio_net_device *vndev,
 		return -ENOSPC;
 	}
 
-	/**
-	 * Saving the buffer information before reserving the header space.
-	 */
-	buf_start = netbuf->data;
-	buf_len = netbuf->len;
-
-	/**
-	 * Retrieve the buffer header length.
-	 */
-	rc = uk_netbuf_header(netbuf, VTNET_HDR_SIZE_PADDED(vndev));
-	if (unlikely(rc != 1)) {
-		uk_pr_err("Failed to allocate space to prepend virtio header\n");
-		return -EINVAL;
-	}
-	rxhdr = netbuf->data;
-
+	/* Reset the scatter gather list */
 	sg = &rxq->sg;
 	uk_sglist_reset(sg);
 
-	/* Appending the header buffer to the sglist */
-	uk_sglist_append(sg, rxhdr, virtio_net_hdr_size(vndev));
+	if (vndev->buf_descr_count == VIRTIO_NET_BUF_DESCR_COUNT_INLINE) {
+		rc = uk_netbuf_header(netbuf, virtio_net_hdr_size(vndev));
+		if (unlikely(rc != 1)) {
+			uk_pr_err("Failed to allocate space to prepend virtio header\n");
+			return -EINVAL;
+		}
+		uk_sglist_append(sg, netbuf->data, netbuf->len);
+	} else {
+		/**
+		 * Saving the buffer information before reserving the header
+		 * space.
+		 */
+		buf_start = netbuf->data;
+		buf_len = netbuf->len;
 
-	/* Appending the data buffer to the sglist */
-	uk_sglist_append(sg, buf_start, buf_len);
+		/**
+		 * Retrieve the buffer header length.
+		 */
+		rc = uk_netbuf_header(netbuf, VTNET_HDR_SIZE_PADDED(vndev));
+		if (unlikely(rc != 1)) {
+			uk_pr_err("Failed to allocate space to prepend virtio header\n");
+			return -EINVAL;
```

Comment:
Why not forward the actual `rc`? :thinking:

Reply:
I haven't written the code. I guess the intention is that if you give it a buffer that returns an error code here, it is an invalid argument error.